### PR TITLE
CSHARP-2483: Support 'startAfter' option to the $changeStream stage.

### PR DESCRIPTION
**Pull request by DmitryLukyanov**

### CODE REVIEW

File: src/MongoDB.Driver.Core/Core/Operations/ChangeStreamOperation.cs [hunk 1]
```diff
@@ -325,6 +346,49 @@ public Task<IAsyncCursor<RawBsonDocument>> ResumeAsync(IReadBinding binding, Can
...
+            // The driver MUST set resumeAfter to the value of the originally used startAfter.
+            // The driver MUST NOT set startAfter. The driver MUST remove it from the original aggregation command.
+            // The driver MUST NOT set startAtOperationTime.
+            if (_startAfter != null)
+            {
+                _resumeAfter = _startAfter;
+                _startAfter = null;
+                _startAtOperationTime = null;
+            }
+            // Docs: if the ChangeStream was created with resumeAfter:
```
Comment (DmitryLukyanov) [inline src/MongoDB.Driver.Core/Core/Operations/ChangeStreamOperation.cs]: this question has been asked in the SPEC ticket: https://jira.mongodb.org/browse/SPEC-1227 partially.
We can have `_resumeAfter` value here which can be originally set in the `_startAfter` field. 
So, I'm not sure that below condition fully satisfies to this: `if the ChangeStream was created with `resumeAfter`:`

File: src/MongoDB.Driver.Core/Core/Operations/ChangeStreamOperation.cs [hunk 2]
```diff
@@ -325,6 +346,49 @@ public Task<IAsyncCursor<RawBsonDocument>> ResumeAsync(IReadBinding binding, Can
...
+            // The driver MUST set resumeAfter to the cached postBatchResumeToken.
+            // The driver MUST NOT set startAfter. If startAfter was in the original aggregation command, the driver MUST remove it.
+            // The driver MUST NOT set startAtOperationTime. If startAtOperationTime was in the original aggregation command, the driver MUST remove it.
+            // todo: should be implemented in scope of the CSHARP-2514.
+
+            // Docs: if the ChangeStream has a cached documentResumeToken:
+            // The driver MUST set resumeAfter to the cached documentResumeToken.
+            // The driver MUST NOT set startAfter. If startAfter was in the original aggregation command, the driver MUST remove it.
+            // The driver MUST NOT set startAtOperationTime. If startAtOperationTime was in the original aggregation command, the driver MUST remove it.
+            // todo:
```
Comment (DmitryLukyanov) [inline src/MongoDB.Driver.Core/Core/Operations/ChangeStreamOperation.cs]: Is this condition the same with the condition in the 378 line below? 
(At least until we will have implemented the ticket 2514).

File: src/MongoDB.Driver.Core/Core/Operations/ChangeStreamOperation.cs [hunk 3]
```diff
@@ -325,6 +346,53 @@ public Task<IAsyncCursor<RawBsonDocument>> ResumeAsync(IReadBinding binding, Can
...
+            // The driver MUST set resumeAfter to the value of the originally used resumeAfter.
+            // The driver MUST NOT set startAfter.
+            // The driver MUST NOT set startAtOperationTime.
+            else if (_resumeAfter != null)
+            {
+                _startAfter = null;
+                _startAtOperationTime = null;
+            }
+
+            // Docs: if the ChangeStream has a saved operation time (either from an originally
```
Comment (DmitryLukyanov) [inline src/MongoDB.Driver.Core/Core/Operations/ChangeStreamOperation.cs]: I'm not sure that below conditions should be implemented in this ticket.

File: src/MongoDB.Driver/ChangeStreamOptions.cs [hunk 4]
```diff
@@ -26,11 +26,7 @@ public class ChangeStreamOptions
     {
         // private fields
         private int? _batchSize;
-        private Collation _collation;
-        private ChangeStreamFullDocumentOption _fullDocument = ChangeStreamFullDocumentOption.Default;
         private TimeSpan? _maxAwaitTime;
-        private BsonDocument _resumeAfter;
-        private BsonTimestamp _startAtOperationTime;
```
Comment (rstam) [inline src/MongoDB.Driver/ChangeStreamOptions.cs]: Don't do this. It's fine to stick with explicit backing fields. There is no need to use automatic properties just because they exist.

Definitely let's never use mixed models, like this, where some of the state of the class is held in explicit backing fields and the rest of the state is held in automatic properties. That's just really confusing. 

After this change the risk is that one looks at this class and immediately assumes that the state of this class consists of only two fields: _batchSize and _maxAwaitTime. But that is wrong, the rest of the state now exists in automatic properties which can only be found by visually scanning for them.

Stick to explicit backing fields.

File: src/MongoDB.Driver/ChangeStreamStageOptions.cs [hunk 5]
```diff
@@ -22,58 +22,44 @@ namespace MongoDB.Driver
     /// </summary>
     public class ChangeStreamStageOptions
     {
-        // private fields
-        private bool? _allChangesForCluster;
-        private ChangeStreamFullDocumentOption _fullDocument = ChangeStreamFullDocumentOption.Default;
-        private BsonDocument _resumeAfter;
-        private BsonTimestamp _startAtOperationTime;
```
Comment (rstam) [inline src/MongoDB.Driver/ChangeStreamStageOptions.cs]: Stick to explicit backing fields.

File: src/MongoDB.Driver.Core/Core/Operations/ChangeStreamOperation.cs [hunk 6]
```diff
@@ -325,6 +346,53 @@ public Task<IAsyncCursor<RawBsonDocument>> ResumeAsync(IReadBinding binding, Can
...
+
+            // Docs: if the ChangeStream was created with startAfter:
+            // The driver MUST set resumeAfter to the value of the originally used startAfter.
+            // The driver MUST NOT set startAfter. The driver MUST remove it from the original aggregation command.
+            // The driver MUST NOT set startAtOperationTime.
+            if (_startAfter != null)
+            {
+                _resumeAfter = _startAfter;
+                _startAfter = null;
+                _startAtOperationTime = null;
```
Comment (rstam) [inline src/MongoDB.Driver.Core/Core/Operations/ChangeStreamOperation.cs]: Let's not mutate the state of the class. That has the potential to make debugging harder because it becomes harder to tell whether the state was set by the caller or mutated by this method.

It would be fine to have a method that throws exceptions if incompatible options are set.

And we can have methods that compute (without mutating state) the "effective" values to use when building commands to send to the server.

**Review by rstam — CHANGES_REQUESTED** [review]

**Comment by rstam** [comment]
Minor style related comments in this initial review.

A more thorough review by me is still pending. I have to read the specs a bit first to review whether the logic changes are correct.

File: src/MongoDB.Driver/ChangeStreamOptions.cs (hunk 4 above)
Comment (DmitryLukyanov) [reply src/MongoDB.Driver/ChangeStreamOptions.cs]: To me having backing fields code at least for classes which provide only information about a business object (without business logic) makes the code unnecessarily bigger and difficult.

But Ok, I will change it.

File: src/MongoDB.Driver.Core/Core/Operations/ChangeStreamOperation.cs (hunk 6 above)
Comment (DmitryLukyanov) [reply src/MongoDB.Driver.Core/Core/Operations/ChangeStreamOperation.cs]: As I understand, it's a requirement from the spec. 
>It would be fine to have a method that throws exceptions if incompatible options are set.

In spec, there is a requirement that we should not throw an exception in case if  incompatible options are set.

>And we can have methods that compute (without mutating state) the "effective" values to use when building commands to send to the server.

Do you mean to have a method which will take `ChangeStreamOperation` as an income parameter and return the modified object?

Comment (rstam) [reply src/MongoDB.Driver.Core/Core/Operations/ChangeStreamOperation.cs]: No. I just don't want to see the fields of the object mutated.

We can compute the effective resumeAfter or startAfter (etc..) instead of mutating the fields of the class.

File: src/MongoDB.Driver/ChangeStreamOptions.cs (hunk 4 above)
Comment (rstam) [reply src/MongoDB.Driver/ChangeStreamOptions.cs]: I'm sure reasonable people can have different opinions on when to use explicit backing fields or not.

But please, let's not MIX explicit backing fields and automatic properties in the same class. That's the worst possible combination.

**Review by rstam — CHANGES_REQUESTED** [review]

File: src/MongoDB.Driver.Core/Core/Operations/ChangeStreamOperation.cs (hunk 2 above)
Comment (rstam) [reply src/MongoDB.Driver.Core/Core/Operations/ChangeStreamOperation.cs]: I don't understand the question. Given that this method should go away anyway I would rephrase the question after refactoring.

File: src/MongoDB.Driver.Core/Core/Operations/ChangeStreamOperation.cs (hunk 1 above)
Comment (rstam) [reply src/MongoDB.Driver.Core/Core/Operations/ChangeStreamOperation.cs]: It's hard to talk about "was created with" if we are mutating the object.

If we had an object we don't mutate the the value of the field IS the original value.

File: src/MongoDB.Driver.Core/Core/Operations/ChangeStreamOperation.cs (hunk 3 above)
Comment (rstam) [reply src/MongoDB.Driver.Core/Core/Operations/ChangeStreamOperation.cs]: We should send the correct values for resumeAfter, startAfter and startAtOperationTime to the server.

Perhaps it would be easier to also implement support for postBatchResumeToken at the same time in order to not have to figure out how to implement "half" of the resume logic.

**Review by rstam — CHANGES_REQUESTED** [review]

**Comment by DmitryLukyanov** [comment]
Evergreen:  https://evergreen.mongodb.com/version/5c8a5d151e2d1702d1faed99

NOTE:
1. There are new tests in spec folder for change stream logic, mostly about some new `operationType`s, but I didn't find any mentioning of `startAfter` options inside them, so I didn't include them to pushed files. I can add them to a solution (I've almost finished it), but It looks like it should be in the scope of another ticket.
2. The current implementation doesn't contain processing of `postBatchResumeToken`. It should be done in the scope of the `CSHARP-2514`.

File: src/MongoDB.Driver.Core/Core/Operations/ChangeStreamCursor.cs [hunk 7]
```diff
@@ -155,7 +154,17 @@ private IEnumerable<TDocument> DeserializeDocuments(IEnumerable<RawBsonDocument>
 
             if (lastRawDocument != null)
             {
-                _changeStreamOperation.ResumeAfter = lastRawDocument["_id"].DeepClone().AsBsonDocument;
+                var resumeToken = lastRawDocument["_id"].DeepClone().AsBsonDocument;
+
+                if (_changeStreamOperation.StartAfter != null && _changeStreamOperation.ResumeAfter == null)
```
Comment (DmitryLukyanov) [inline src/MongoDB.Driver.Core/Core/Operations/ChangeStreamCursor.cs]: I've changed this logic. In the previous implementation, I just set `StartAfter` to `null`.
But it looks like here we can reinitialize field which was originally initialized when we created `watch`

File: src/MongoDB.Driver.Core/Core/Operations/ChangeStreamOperation.cs [hunk 8]
```diff
@@ -85,7 +93,9 @@ public class ChangeStreamOperation<TResult> : IChangeStreamOperation<TResult>
         private readonly IReadOnlyList<BsonDocument> _pipeline;
         private ReadConcern _readConcern = ReadConcern.Default;
         private readonly IBsonSerializer<TResult> _resultSerializer;
+
```
Comment (rstam) [inline src/MongoDB.Driver.Core/Core/Operations/ChangeStreamOperation.cs]: Unnecessary blank line.

File: src/MongoDB.Driver.Core/Core/Operations/ChangeStreamCursor.cs [hunk 9]
```diff
@@ -155,7 +154,17 @@ private IEnumerable<TDocument> DeserializeDocuments(IEnumerable<RawBsonDocument>
...
+                {
+                    _changeStreamOperation.StartAfter = resumeToken;
+                }
+                else
+                {
+                    _changeStreamOperation.ResumeAfter = resumeToken;
+                }
+
```
Comment (rstam) [inline src/MongoDB.Driver.Core/Core/Operations/ChangeStreamCursor.cs]: As I mentioned in an earlier code review I would like us to NOT alter any fields originally set by the creator of the operation. Instead, just store the resume token in a new field:

    if (lastRawDocument != null)
    {
        _changeStreamOperation.DocumentResumeToken = lastRawDocument["_id"].DeepClone().AsBsonDocument;
    }

File: src/MongoDB.Driver.Core/Core/Operations/ChangeStreamOperation.cs [hunk 10]
```diff
@@ -85,7 +93,9 @@ public class ChangeStreamOperation<TResult> : IChangeStreamOperation<TResult>
...
         private BsonDocument _resumeAfter;
+        private BsonDocument _startAfter;
         private BsonTimestamp _startAtOperationTime;
```
Comment (rstam) [inline src/MongoDB.Driver.Core/Core/Operations/ChangeStreamOperation.cs:100]: Add a few new fields:

    private BsonDocument _documentResumeToken;
    private BsonTimestamp _initialOperationTime;

The first tracks the most recently seen document resume token.

The second is used (if required) to track the operation time of the initial aggregate command.

File: src/MongoDB.Driver.Core/Core/Operations/ChangeStreamOperation.cs [hunk 11]
```diff
@@ -42,6 +42,14 @@ public interface IChangeStreamOperation<TResult> : IReadOperation<IAsyncCursor<T
...
         BsonDocument ResumeAfter { get; set; }
 
+        /// <summary>
+        /// Gets or sets the start after.
+        /// </summary>
+        /// <value>
+        /// The start after.
+        /// </value>>
+        BsonDocument StartAfter { get; set; }
+
```
Comment (rstam) [inline src/MongoDB.Driver.Core/Core/Operations/ChangeStreamOperation.cs]: Add a new property for the cursor to use to record the most recently seen document result token:

    BsonDocument DocumentResumeToken { get; set; }

Add it in the right place, I couldn't comment on a line that didn't change...

Comment (rstam) [reply src/MongoDB.Driver.Core/Core/Operations/ChangeStreamOperation.cs]: Once DocumentResumeToken has been added we no longer need the new StartAfter property, so remove it.

We also no longer need the existing ResumeAfter and StartAtOperationTime properties. Not sure if we should remove them since they are a part of the public API. At least deprecate them?

File: src/MongoDB.Driver.Core/Core/Operations/ChangeStreamOperation.cs [hunk 12]
```diff
@@ -270,7 +287,7 @@ public IAsyncCursor<TResult> Execute(IReadBinding binding, CancellationToken can
             using (var channelBinding = new ChannelReadBinding(channelSource.Server, channel, binding.ReadPreference, binding.Session.Fork()))
             {
                 cursor = Resume(channelBinding, cancellationToken);
-                if (_startAtOperationTime == null && _resumeAfter == null)
+                if (_startAtOperationTime == null && _resumeAfter == null && _startAfter == null)
```
Comment (rstam) [inline src/MongoDB.Driver.Core/Core/Operations/ChangeStreamOperation.cs]: Add a new helper method ExecuteAggregateOperation that takes a resuming parameter indicating whether we are executing the first aggregate command or one to resume the change stream. Also, refactor lines 290-297 to a helper method.

So lines 289-297 become:

    cursor = ExecuteAggregateOperation(channelBinding, resuming: false, cancellationToken);
    SaveInitialOperationTimeIfRequired(channel, channelBinding, cursor);

Comment (rstam) [reply src/MongoDB.Driver.Core/Core/Operations/ChangeStreamOperation.cs]: The new helper method would be:

    private void SaveInitialOperationTimeIfRequired(IChannelHandle channel, IReadBinding binding, IAsyncCursor<RawBsonDocument> cursor)
    {
        if (_startAtOperationTime == null && _resumeAfter == null && _startAfter == null)
        {
            var maxWireVersion = channel.ConnectionDescription.IsMasterResult.MaxWireVersion;
            if (maxWireVersion >= 7)
            {
                if (_documentResumeToken != null)
                {
                    _initialOperationTime = binding.Session.OperationTime;
                }
            }
        }
    }

Note the *new* check that _documentResumeToken != null. That's from the spec.

File: src/MongoDB.Driver.Core/Core/Operations/ChangeStreamOperation.cs [hunk 13]
```diff
@@ -298,7 +315,7 @@ public async Task<IAsyncCursor<TResult>> ExecuteAsync(IReadBinding binding, Canc
             using (var channelBinding = new ChannelReadBinding(channelSource.Server, channel, binding.ReadPreference, binding.Session.Fork()))
             {
                 cursor = await ResumeAsync(channelBinding, cancellationToken).ConfigureAwait(false);
-                if (_startAtOperationTime == null && _resumeAfter == null)
+                if (_startAtOperationTime == null && _resumeAfter == null && _startAfter == null)
```
Comment (rstam) [inline src/MongoDB.Driver.Core/Core/Operations/ChangeStreamOperation.cs]: Similar to above:

    cursor = await ExecuteAggregateOperationAsync(channelBinding, resuming: false, cancellationToken).ConfigureAwait(false);
    SaveInitialOperationTimeIfRequired(channel, channelBinding, cursor);

File: src/MongoDB.Driver.Core/Core/Operations/ChangeStreamOperation.cs [hunk 14]
```diff
@@ -356,6 +373,7 @@ private BsonDocument CreateChangeStreamStage()
             {
                 { "fullDocument", ToString(_fullDocument) },
                 { "allChangesForCluster", true, _collectionNamespace == null && _databaseNamespace == null },
+                { "startAfter", _startAfter, _startAfter != null},
```
Comment (rstam) [inline src/MongoDB.Driver.Core/Core/Operations/ChangeStreamOperation.cs]: CreateChangeStreamStage should be refactored to consider whether it is a resuming operation and to use the "effective" resume/start values:

    private BsonDocument CreateChangeStreamStage(bool resuming)
    {
        var resumeStartValues = GetEffectiveResumeStartValues(resuming);

        var changeStreamOptions = new BsonDocument
        {
            { "fullDocument", ToString(_fullDocument) },
            { "allChangesForCluster", true, _collectionNamespace == null && _databaseNamespace == null },
            { "startAfter", resumeStartValues.StartAfter, resumeStartValues.StartAfter != null},
            { "startAtOperationTime", resumeStartValues.StartAtOperationTime, resumeStartValues.StartAtOperationTime != null },
            { "resumeAfter", resumeStartValues.ResumeAfter, resumeStartValues.ResumeAfter != null }
        };
        return new BsonDocument("$changeStream", changeStreamOptions);
    }

It is this approach of "effective resume/start values" that allows us to NOT overwrite the original values supplied by the creator of the operation.

Comment (rstam) [reply src/MongoDB.Driver.Core/Core/Operations/ChangeStreamOperation.cs]: The GetEffectiveResumeStartValues should look like:

    private ResumeStartValue GetEffectiveResumeStartValues(bool resuming)
    {
        if (resuming)
        {
            if (_documentResumeToken != null)
            {
                return new ResumeStartValue { ResumeAfter = _documentResumeToken };
            }

            if (_startAfter != null)
            {
                return new ResumeStartValue { ResumeAfter = _startAfter };
            }

            if (_resumeAfter != null)
            {
                return new ResumeStartValue { ResumeAfter = _resumeAfter };
            }

            if (_startAtOperationTime != null || _initialOperationTime != null)
            {
                return new ResumeStartValue { StartAtOperationTime = _startAtOperationTime ?? _initialOperationTime };
            }
        }

        return new ResumeStartValue { ResumeAfter = _resumeAfter, StartAfter = _startAfter, StartAtOperationTime = _startAtOperationTime };
    }

NOTE: by refactoring to not change the original _startAfter etc.. values supplied by the creator of the operation this code now reads almost exactly like the "Resume Process" section of the spec (minus the postBatchResumeToken which will be added later).

The return value is a simple DTO:

    private struct ResumeStartValue
    {
        public BsonDocument ResumeAfter { get; set; }
        public BsonDocument StartAfter { get; set; }
        public BsonTimestamp StartAtOperationTime { get; set; }
    }

**Review by rstam — CHANGES_REQUESTED** [review]
The main change I am requesting is that we NOT alter the resumeAfter/startAfter/startAtOperationTime fields originally set by the creator of the ChangeStreamOperation.

I realize we were already doing that... but not doing that allows us to write code that reads almost like the spec.

**Comment by rstam** [comment]
There were a few comments I couldn't add because github doesn't support writing comments on lines that haven't changed.

I've prototyped my suggestions here:

https://github.com/rstam/mongo-csharp-driver/commit/ba1f99157fae791f6f857dce93a8fb7c471613ad

File: src/MongoDB.Driver.Core/Core/Operations/ChangeStreamOperation.cs (hunk 8 above)
Comment (DmitryLukyanov) [reply src/MongoDB.Driver.Core/Core/Operations/ChangeStreamOperation.cs]: Done

File: src/MongoDB.Driver.Core/Core/Operations/ChangeStreamCursor.cs (hunk 9 above)
Comment (DmitryLukyanov) [reply src/MongoDB.Driver.Core/Core/Operations/ChangeStreamCursor.cs]: Done

File: src/MongoDB.Driver.Core/Core/Operations/ChangeStreamOperation.cs (hunk 10 above)
Comment (DmitryLukyanov) [reply src/MongoDB.Driver.Core/Core/Operations/ChangeStreamOperation.cs:100]: Done

File: src/MongoDB.Driver.Core/Core/Operations/ChangeStreamOperation.cs (hunk 11 above)
Comment (DmitryLukyanov) [reply src/MongoDB.Driver.Core/Core/Operations/ChangeStreamOperation.cs]: Done

File: src/MongoDB.Driver.Core/Core/Operations/ChangeStreamOperation.cs (hunk 12 above)
Comment (DmitryLukyanov) [reply src/MongoDB.Driver.Core/Core/Operations/ChangeStreamOperation.cs]: Done

File: src/MongoDB.Driver.Core/Core/Operations/ChangeStreamOperation.cs (hunk 13 above)
Comment (DmitryLukyanov) [reply src/MongoDB.Driver.Core/Core/Operations/ChangeStreamOperation.cs]: Done

File: src/MongoDB.Driver.Core/Core/Operations/ChangeStreamOperation.cs (hunk 14 above)
Comment (DmitryLukyanov) [reply src/MongoDB.Driver.Core/Core/Operations/ChangeStreamOperation.cs]: Done

File: src/MongoDB.Driver.Core/Core/Operations/ChangeStreamOperation.cs (hunk 11 above)
Comment (rstam) [reply src/MongoDB.Driver.Core/Core/Operations/ChangeStreamOperation.cs]: You still need to remove this new property from the interface.

We're not removing any existing properties, but the only new property we are adding is DocumentResumeToken.

**Review by rstam — CHANGES_REQUESTED** [review]
Looks like one minor change got missed.

**Review by rstam — APPROVED** [review]
Looks good!
